### PR TITLE
Disallow users from updating selector or template labels in kubectl apply

### PR DESCRIPTION
**Pull request by janetkuo**
**What this PR does / why we need it**: Ref https://github.com/kubernetes/kubernetes/issues/26202#issuecomment-245114043 -- updating selectors / pod template labels is as dangerous as updating name, and should be disallowed. In this PR we only disallow this in `kubectl apply`, and we should do more in the future PRs. 

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #

**Special notes for your reviewer**:

**Release note**:

``` release-note
```

@kubernetes/kubectl @kubernetes/sig-apps @mwielgus @pwittrock @bgrant0607 @ghodss @AdoHe @lavalamp @erictune 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32228)

### CODE REVIEW

**Comment by janetkuo** [comment]
This breaks backward compatibility. Should we add a flag so that people who know what they're doing can still update selector / pod template labels?

**Comment by janetkuo** [comment]
Should we say something like "create a new resource instead" in the error message, or is it redundant?

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -274,8 +279,39 @@ func (p *patcher) patch(current runtime.Object, modified []byte, source, namespa
...
+
+// checkIllegalChanges returns error when illegal updates are made, right now only
+// pod template labels updates and selector updates are disallowed.
+// TODO: do this in other kubectl update commands or move this to server-side
+func checkIllegalChanges(patch []byte, name, resourceType string) error {
+	diff := make(map[string]interface{})
+	if err := json.Unmarshal(patch, &diff); err != nil {
+		return fmt.Errorf("failed to unmarshal patch when determining if the change is legal: %v", err)
+	}
+	templateLabelsDiff := getNestedField(diff, "spec", "template", "metadata", "labels")
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/apply.go]: At minimum move this to cmdutil.Factory and pass in the info.Mapping for the object.

**Comment by AdoHe** [comment]
generally lgtm. or we can `disable-illegal-check` to disable this check.

**Comment by smarterclayton** [comment]
This needs to be for known types only.  Just because something has spec.templates.metadata.labels doesn't mean it's unsafe to change selectors.  Looking at any object and applying structural rules is not safe.

**Comment by janetkuo** [comment]
@smarterclayton What if I look at `info.Mapping.Resource` and only do this check for certain resources, such as deployments, replication controllers, etc.?

**Comment by smarterclayton** [comment]
Yeah that's fine.  Just needs to be a known list in Factory (so that
it can be extended to other compiled in resources).

**Pull request closed without merging**